### PR TITLE
Adds a proof harness for the IotMqtt_Init function

### PR DESCRIPTION
**Pull request by feliperodri**
*Description of changes:*

This PR depends on https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/780!

- Adds `.gitignore` file on cbmc directory;
- Updates CBMC flags in the `Makefile.common`;
- Adds a new proof harness for the `IotMqtt_Init` function.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+#include <stdlib.h>
+#include <string.h>
+
+void harness()
+{
+  IotMqtt_Init();
+}
+
+
```
Comment (gordonwang0) [inline cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c]: Unnecessary whitespace at the end of this file.

File: cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#include "iot_config.h"
+#include "private/iot_mqtt_internal.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+void harness()
+{
+  IotMqtt_Init();
```
Comment (danielsn) [inline cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c]: I assume this initializes a global structure? Did we make sure to run with `--nondet-static` or whatever the flag is?

File: cbmc/proofs/IotMqtt_Init/cbmc-batch.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+jobos: ubuntu16
+cbmcflags:        '=--unwind;1;--bounds-check;--pointer-check;--unwinding-assertions;--nondet-static='
+goto: IotMqtt_Init.goto
+expected: SUCCESSFUL
```
Comment (danielsn) [inline cbmc/proofs/IotMqtt_Init/cbmc-batch.yaml]: alphabetize the fields.

File: cbmc/proofs/IotMqtt_Init/Makefile [hunk 4]
```diff
@@ -0,0 +1,30 @@
+ENTRY=IotMqtt_Init
+
+ABSTRACTIONS += --remove-function-body IotLog_Generic
+ABSTRACTIONS += --remove-function-body Iot_EnterCritical
+ABSTRACTIONS += --remove-function-body Iot_ExitCritical
```
Comment (danielsn) [inline cbmc/proofs/IotMqtt_Init/Makefile]: Alphabatize

File: cbmc/proofs/Makefile.common [hunk 5]
```diff
@@ -14,6 +14,9 @@ GOTO_ANALYZER ?= goto-analyzer
 BATCH ?= cbmc-batch
 VIEWER ?= cbmc-viewer
 
+CBMC_OBJECT_BITS = 8
```
Comment (danielsn) [inline cbmc/proofs/Makefile.common]: Take a look at https://github.com/awslabs/s2n/blob/master/tests/cbmc/proofs/Makefile.common

File: cbmc/proofs/Makefile.common [hunk 6]
```diff
@@ -62,11 +64,24 @@ $(ENTRY).goto: $(ENTRY)2.goto
...
+	--nondet-static \
+
+CBMCFLAGS += \
+	--div-by-zero-check \
+	--float-overflow-check \
+	--nan-check \
+	--pointer-overflow-check \
+	--undefined-shift-check \
+	--signed-overflow-check \
+	--unsigned-overflow-check \
```
Comment (danielsn) [inline cbmc/proofs/Makefile.common]: Why is this split into three sections?

File: cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c (hunk 2 above)
Comment (feliperodri) [reply cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c]: Yes, take a look at `cbmc-batch.yaml `, it adds the `--nondet-static` flag.

[the remark this answers is not in the document]
Comment (feliperodri) [reply cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c]: We are not initializing a structure here, but I added a check on whether the status output is as expected.

File: cbmc/proofs/IotMqtt_Init/Makefile [hunk 7]
```diff
@@ -0,0 +1,30 @@
+ENTRY=IotMqtt_Init
+
+ABSTRACTIONS += --remove-function-body IotLog_Generic
```
[the remark this answers is not in the document]
Comment (feliperodri) [reply cbmc/proofs/IotMqtt_Init/Makefile]: Done!

File: cbmc/proofs/IotMqtt_Init/cbmc-batch.yaml [hunk 8]
```diff
@@ -0,0 +1,4 @@
+jobos: ubuntu16
+cbmcflags:        '=--unwind;1;--bounds-check;--pointer-check;--unwinding-assertions;--nondet-static='
```
[the remark this answers is not in the document]
Comment (feliperodri) [reply cbmc/proofs/IotMqtt_Init/cbmc-batch.yaml]: Done!

File: cbmc/proofs/IotMqtt_Init/Makefile (hunk 4 above)
Comment (feliperodri) [reply cbmc/proofs/IotMqtt_Init/Makefile]: Done!

File: cbmc/proofs/Makefile.common (hunk 6 above)
Comment (feliperodri) [reply cbmc/proofs/Makefile.common]: Merged it into one.

File: cbmc/proofs/IotMqtt_Init/Makefile [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+UNWINDING += --unwind 1
+
+default: report
+
+# Use the generic implementations of atomic operations.
+# Using the gcc atomic intrinsics causes coverage computation to fail.
+# Define the generic flag here, and replace *_gcc.h with *_generic.h below.
+DEF += -DIOT_ATOMIC_GENERIC=1
+
+install_atomic:
```
[the remark this answers is not in the document]
Comment (feliperodri) [reply cbmc/proofs/IotMqtt_Init/Makefile]: Done!

File: cbmc/proofs/IotMqtt_Init/Makefile [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ABSTRACTIONS += --remove-function-body Iot_ExitCritical
+
+OBJS += $(ENTRY)_harness.goto
+OBJS += $(MQTT)/libraries/standard/mqtt/src/iot_mqtt_api.goto
+
+UNWINDING += --unwind 1
+
+default: report
+
+# Use the generic implementations of atomic operations.
```
[the remark this answers is not in the document]
Comment (feliperodri) [reply cbmc/proofs/IotMqtt_Init/Makefile]: Not a problem anymore.

File: cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c [hunk 11]
```diff
@@ -0,0 +1,11 @@
+#include "iot_config.h"
+#include "private/iot_mqtt_internal.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+void harness()
+{
+  IotMqttError_t status = IotMqtt_Init();
+  assert(status == IOT_MQTT_SUCCESS || status == IOT_MQTT_NOT_INITIALIZED);
```
Comment (danielsn) [inline cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c]: Don't you have an `is_valid_mqtt` or anything you can assert here?

Comment (markrtuttle) [reply cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c]: I can't think of another place in the project where we would use a predicate like this.

File: cbmc/.gitignore [hunk 12]
```diff
@@ -0,0 +1,6 @@
+proofs/ninja.py
```
Comment (markrtuttle) [inline cbmc/.gitignore]: Is it good etiquette to add to the .gitignore something that is our own development tool?  This is actually kind of dangerous, because I routinely do "git clean -fdx" to blow away all my trash, and this will remove the development tools listed in gitignore.

File: cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c (hunk 11 above)
Comment (feliperodri) [reply cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c]: @danielsn Basically, `IotMqtt_Init`sets `_initCalled != MQTT_LIBRARY_UNINITIALIZED`. Thus, we could check here the following:
```c
assert(_checkInit()); // checks if _initCalled != MQTT_LIBRARY_UNINITIALIZED
```
However, the problem is that the `_initCalled` is volatile, so CBMC will *always* treat it as non-deterministic. See,
```c
static volatile uint32_t _initCalled = MQTT_LIBRARY_UNINITIALIZED;
```
I'd keep the proof harness as it is until we have another approach to treat volatile variables on CBMC.

File: cbmc/.gitignore (hunk 12 above)
Comment (feliperodri) [reply cbmc/.gitignore]: Done!

File: cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c (hunk 11 above)
Comment (markrtuttle) [reply cbmc/proofs/IotMqtt_Init/IotMqtt_Init_harness.c]: I think this is fine.
